test/hotspot/jtreg/serviceability/tmtools/jstat/utils/GcProvoker.java

```diff
@@ -58,17 +58,33 @@ private List<Object> allocateHeap(float targetUsage) {
      * This method provokes a GC
      */
     public void provokeGc() {
+        float targetFraction = 0;
+        // Read sizes of eden and overall heap, to find eden size as a fraction of heap.
+        // Recognise if heap is changing size, and retry (heap is expected to initially shrink).
         for (int i = 0; i < 3; i++) {
+            // One retry is expected occasionally.  More "should not happen":
+            if (i == 2) {
+                throw new RuntimeException("Cannot calculate targetFraction.  Heap size not stable.");
+            }
+            long heapSize0 = Pools.getHeapCommittedSize();
             long edenSize = Pools.getEdenCommittedSize();
             long heapSize = Pools.getHeapCommittedSize();
-            float targetPercent = ((float) edenSize) / (heapSize);
-            if ((targetPercent < 0) || (targetPercent > 1.0)) {
-                throw new RuntimeException("Error in the percent calculation" + " (eden size: " + edenSize + ", heap size: " + heapSize + ", calculated eden percent: " + targetPercent + ")");
+            if (heapSize < heapSize0) {
+                System.out.println("provokeGc: Heap shrinking, retry. eden: " + edenSize + ", heap0: " + heapSize0 + ", heap: " + heapSize);
+                System.gc();
+                continue;
+            } else {
```

Comment:
In fact, the `else-statement` is not needed here because the `if-statement` is ended with `continue`.
At least, this looks confusing to me.

Reply:
OK sure, just updated to remove else, and add a comment which may help.